CIPs/cip-0032.md

```diff
@@ -13,81 +13,105 @@ license: Apache 2.0
+
+- `TargetCompletionRate` - The target completion rate for Attestation Services, taking into account standard user drop-off and other factors outside of the AS control.
+- `AttestationServiceRewardPercentage` - The percentage of max rewards that are allocated to running an Attestation Service. Initially `.20`.
+- `MinGlobalAttestationCompletionRate` - The minimum completion rate across all validators for rewards to be reduced for registered Attestation Services.
+- `GlobalAttestationFailureFallbackReward` - The rate that Attestation Service operators will be rewarded when the global completion rate drops below `MinGlobalAttestationCompletionRate`.
 
-- `MinAttestationSlashablePeriod` is the number of consecutive blocks in which an elected Attestation Service is required to complete at least 1 attestation.
-- `AttestationSlasherPenalty` is the penalty amount that is slashed in a slashing event.
-- `AttestationSlasherReward` is the reward amount that a slasher is granted.
-- `MinAttestationSlashablePeriod`, `AttestationSlasherPenalty`, and `AttestationSlasherReward` are easily configurable via governance proposals.
+- `AttestationMinPeriod` - The period which the issuer must complete at least one successful attestation or potentially face deregistration.
+- `AttestationMinRequested` - The minimum number of attestations requested from the issuer to be eligible for removal.
+- `MinNumberOfAttestationIssuers` - If deregistering an Attestation Service would drop the count below this number, then prevent removal.
 
-This proposal introduces a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `MinAttestationSlashablePeriod` blocks. The `Attestations` contract will be upgraded to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+The `Attestations` contract will be upgraded to register the request and completion events. It will be read by the new `Validators` contract. The `Attestations` contract will contain the following state to track the request and completion counts per epoch:
 
 ```js
-// Tracks the last time an attestation was requested from an issuer (AS node)
-mapping(address => uint256) public lastAttestationRequested;
+// Tracks per issuer completion rate
+// These values are reset per issuer from Validators.distributeEpochPaymentsFromSigner()
+mapping(address => uint256) public attestationRequestCount;
+mapping(address => uint256) public attestationCompletionCount;
+
+// Track global completion rate
+// currEpochIndex = getEpochNumberOfBlock(block.number) % 2
+// These values get reset when epoch changes (during the first attestation request)
+uint256 public epochRequestCount;
+uint256 public epochIndexToCompletionCount;
+uint8 private currEpochIndex;
```

Comment:
Because it is not grouped with other state variables, this variable will still take up a full 256-bit word. So unless there is a natural grouping, we might as well make this `uint256`. (We could certainly turn `epochRequestCount` and `epochIndexToCompletionCount` into `uint64` though.)

Reply:
I was thinking the same.. I left it as `uint8` in case we decide to add more state that is sub-256. Will leave it up to the implementer to decide.

Reply:
Good point on the other vars. I'll update them.